Sources/Signal.swift

```diff
@@ -159,9 +153,13 @@ public final class Signal<Value, Error: Swift.Error> {
 	}
 
 	deinit {
-		if state.swap(nil) != nil {
-			// As the signal can deinitialize only when it has no observers attached,
-			// only the generator disposable has to be disposed of at this point.
+		// A signal can deinitialize only when it is not retained and has no
+		// active observers. So `state` need not be swapped.
+		//
+		// The generator disposable needs to be disposed of only if a `green` status
+		// is observed here, since `red` would mean it has already been disposed of,
+		// and `yellow` is impossible to be observed here.
+		if case .green = status {
```

Comment:
It is now swapped out before disposal in https://github.com/ReactiveCocoa/ReactiveSwift/pull/137/commits/63a7e0d28db724191a6a2254458f5dbe91f0f196, so no check is needed.